### PR TITLE
fix (razzle #1704) by adding support for webpack-dev-server v4

### LINKED ISSUE #1704: Allow Razzle 4 to use webpack-dev-server v4.0
## 🚀 Feature request

Add an option that will allow me to use `webpack-dev-server` v4. We are using a security code monitor and it detected a vulnerability with v3 that is fixed in v4.

Warning from the security monitor tool:

<h5 style="box-sizing: border-box; margin-bottom: 16px; margin-top: 24px; font-size: 0.875em; font-weight: 600; line-height: 1.25; color: rgb(36, 41, 47); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">With an upgrade:</h5>

Severity | Priority Score (*) | Issue | Breaking Change | Exploit Maturity
-- | -- | -- | -- | --
 Medium | 586/1000Why? Proof of Concept exploit, Has a fix available, CVSS 5.3 | Regular Expression Denial of Service [(ReDoS)SNYK-JS-GLOBPARENT-1016905](https://snyk.io/vuln/SNYK-JS-GLOBPARENT-1016905) | Yes | Proof of Concept

### Current Behavior

The generated webpack config has incompatible `devServer` options for `webpack-dev-server` since they made breaking configuration changes.

```
ValidationError: Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'before'. These properties are valid:
   object { allowedHosts?, bonjour?, client?, compress?, devMiddleware?, headers?, historyApiFallback?, host?, hot?, http2?, https?, ipc?, liveReload?, onAfterSetupMiddleware?, onBeforeSetupMiddleware?, onListening?, open?, port?, proxy?, setupExitSignals?, static?, watchFiles?, webSocketServer? }
```

### Desired Behavior

I can specify in the razzle options that I'm using `devServerVersion: 4` and it will generate the appropriate `devServer` options for v4 rather than v3.

### Suggested Solution

The official [migration guide](https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md) pretty much shows what is needed to be changes. Perhaps adding to the options list for razzle to indicate the `devServer` version and then changing the options accordingly?

### Who does this impact? Who is this for?

Any user who wants to pick up the latest security fixes for webpack and its associated tools

### Describe alternatives you've considered

I could manually upgrade the `devServer` options provided  in my `razzle.config.js` except that one of the broken options is a callback that I cannot convert without a lot of trial and error.

### PR DESCRIPTION
**Pull request by heath-freenome**
Fixes #1704 

- Updated `package.meta.json` and `packages/razzle-dev-utils/package.json` to allow for v3 and v4 of `webpack-dev-server`
- Updated `createConfigAsync.js` to build the correct `devServer` config for v3 or v4 based on the extracted version from `webpack-dev-server`'s version from the `package.json`
- Added a new `with-webpack-dev-server-v4` example

### CODE REVIEW

**Comment by heath-freenome** [comment]
@fivethreeo Here you go!

File: packages/razzle/config/createConfigAsync.js [hunk 1]
```diff
@@ -834,33 +834,68 @@ module.exports = (
           };
         }
 
+        let devServer;
+        if (razzleOptions.enableDevServerV4) {
```
Comment (fivethreeo) [inline packages/razzle/config/createConfigAsync.js]: Maybe use?

```js
const devserverVersion = require("webpack-dev-server/package.json").version;
console.log(devserverVersion);
```

**Review by fivethreeo — CHANGES_REQUESTED** [review]
Maybe use?

```js
const devserverVersion = require("webpack-dev-server/package.json").version;
console.log(devserverVersion);
```

File: packages/razzle/config/createConfigAsync.js (hunk 1 above)
Comment (heath-freenome) [reply packages/razzle/config/createConfigAsync.js]: @fivethreeo Ok, after a little trial and error, I went with this approach. Let me know how it looks now

File: packages/razzle/config/createConfigAsync.js [hunk 2]
```diff
@@ -834,33 +834,73 @@ module.exports = (
           };
         }
 
+        let devServer;
+        // Fetch the package details from the `webpack-dev-server`
+        const devserverPkg = require('webpack-dev-server/package.json');
```
Comment (fivethreeo) [inline packages/razzle/config/createConfigAsync.js]: Move to top of file so it won't compute every time.

File: packages/razzle/config/createConfigAsync.js [hunk 3]
```diff
@@ -834,33 +834,73 @@ module.exports = (
...
+        // Parse the first character from the `x.y.z` version notation (i.e. major version)
+        const devServerMajorVersion = parseInt(devserverPkg.version[0]);
+        // If the major version is > 3, then use the newer configuration notation
+        if (devServerMajorVersion > 3) {
+          // See https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md for how this was migrated
+          devServer = {
```
Comment (fivethreeo) [inline packages/razzle/config/createConfigAsync.js]: Maybe sett 3 and 4 common options and do special cases for each version, use Object.assign?

File: packages/razzle/config/createConfigAsync.js (hunk 2 above)
Comment (heath-freenome) [reply packages/razzle/config/createConfigAsync.js]: I just moved the require

File: packages/razzle/config/createConfigAsync.js (hunk 3 above)
Comment (heath-freenome) [reply packages/razzle/config/createConfigAsync.js]: Done

File: packages/razzle/config/createConfigAsync.js [hunk 4]
```diff
@@ -847,20 +848,50 @@ module.exports = (
...
-          // Reportedly, this avoids CPU overload on some systems.
-          // https://github.com/facebookincubator/create-react-app/issues/293
-          watchOptions: { ignored: /node_modules/ },
-          before(app) {
-            // This lets us open files from the runtime error overlay.
-            app.use(errorOverlayMiddleware());
-          },
         };
+        // Parse the first character from the `x.y.z` version notation (i.e. major version)
+        const devServerMajorVersion = parseInt(devserverPkg.version[0]);
```
Comment (fivethreeo) [inline packages/razzle/config/createConfigAsync.js]: Can move this one up too.

Comment (heath-freenome) [reply packages/razzle/config/createConfigAsync.js]: Done

**Comment by heath-freenome** [comment]
@fivethreeo when will this fix be available in a razzle release?

**Comment by fivethreeo** [comment]
now :)

**Comment by chrisdostert** [comment]
@fivethreeo It looks like with webpack-dev-server 4, the module casing changed from `webpack-dev-server/client/utils/createSocketUrl` to `webpack-dev-server/client/utils/createSocketURL` at https://github.com/jaredpalmer/razzle/blob/776725f862bb996210e2c69631dc5233c9b2b873/packages/razzle-dev-utils/webpackHotDevClient.js#L15 . Otherwise we get 
```
 WARNING in ./node_modules/webpack-dev-server/client/utils/createSocketURL.js
[1] There are multiple modules with names that only differ in casing.
[1] This can lead to unexpected behavior when compiling on a filesystem with other case-semantic.
[1] Use equal casing. Compare these module identifiers:
[1] * /src/node_modules/webpack-dev-server/client/utils/createSocketURL.js
```

**Comment by fivethreeo** [comment]
Will need to test for version there aswell. But do we really have to swap out the client anymore with newer dev server?

**Comment by fivethreeo** [comment]
released 4.2.1 with module path fix

**Comment by fivethreeo** [comment]
released 4.2.2 with reverted module path fix, issue with filename seems to be a webpack-dev-server issue

**Comment by heath-freenome** [comment]
> released 4.2.2 with reverted module path fix, issue with filename seems to be a webpack-dev-server issue

Is there an issue with `webpack-dev-server` for that?

**Comment by fivethreeo** [comment]
@chrisdostert reported issues with createSocketURL

**Pull request merged**